### PR TITLE
RFC: Added sparse low rank cholesky updates

### PR DESCRIPTION
**Pull request by mfalt**
Implements low rank updates for "Sparse Cholesky"
See issue https://github.com/JuliaLang/LinearAlgebra.jl/issues/6

### CODE REVIEW

File: base/sparse/cholmod.jl [hunk 1]
```diff
@@ -1541,6 +1541,88 @@ ldltfact{T<:Real}(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
+"""
+lowrankupdate{Tv<:VTypes}(F::Factor{Tv}, V::AbstractArray{Tv}) =
+    lowrankupdate!(copy(F), V)
+
+"""
+    lowrankupdate(F::Factor, C) -> FF::Factor
+
+Get an LDLt Factorization of `A+C*C'` given an `LDLt` or `LLt` factorization `F` of `A`.
+
+The returned factor is always an `LDLt` factorization.
```
Comment (andreasnoack) [inline base/sparse/cholmod.jl]: Is this a restriction in CHOLMOD?

**Review by andreasnoack — APPROVED** [review]
Thanks for adding this.

File: base/sparse/cholmod.jl (hunk 1 above)
Comment (mfalt) [inline base/sparse/cholmod.jl]: As far as I understand, yes. See for example [here](https://github.com/PetterS/SuiteSparse/blob/27e5a8516464a6ac40bd3fa0e5b46e51b11f4765/CHOLMOD/Include/cholmod_modify.h#L50).
Maybe there is a way to convert back, but I'm not not sure about the consequences of that.

Comment (andreasnoack) [inline base/sparse/cholmod.jl]: Could you check and confirm what happens with the [is_ll field](https://github.com/JuliaLang/julia/blob/master/base/sparse/cholmod.jl#L278) when updating an `LL'`?

Comment (mfalt) [inline base/sparse/cholmod.jl]: Sure, I just tested and got the expected results:
When called on `LLt` it changes from 1 to 0 
When called on `LDLt` it stays at 0

Comment (mfalt) [inline base/sparse/cholmod.jl]: Actually, I think the reason for conversion is that an LLt factorization will not exist after a downdate to an indefinite matrix. The LDLt can still handle negative elements in D.

Comment (andreasnoack) [inline base/sparse/cholmod.jl]: True, but for the update, it shouldn't be necessary.

**Comment by tkelman** [comment]
Since master is in feature freeze, this will have to wait until after we've branched. Thanks for working on this.

**Comment by tkelman** [comment]
To make these available in existing releases of Julia, it may be worth adding them to https://github.com/JuliaSparse/SuiteSparse.jl

**Comment by andreasnoack** [comment]
Could you please rebase this?

File: base/sparse/cholmod.jl [hunk 2]
```diff
@@ -1541,6 +1541,88 @@ ldltfact{T<:Real}(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
     kws...) = ldltfact(Sparse(A); kws...)
 
+## Rank updates
+
+"""
+    lowrankupdowndate!(F::Factor, C::Sparse, update::Cint)
+
+Update an `LDLt` or `LLt` Factorization `F` of `A` to a factorization of `A±C*C'`.
+
+If sparsity perserving factorization is used, i.e. `L*L'*=P*A*P'` then the new
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: preserving

File: base/sparse/cholmod.jl [hunk 3]
```diff
@@ -1541,6 +1541,88 @@ ldltfact{T<:Real}(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
+factor will be `L*L'=P*A*P'+C'*C`
+
+update: Cint(1) for A+CC', Cint(0) for A-CC'
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: should code highlight the values here

File: base/sparse/cholmod.jl [hunk 4]
```diff
@@ -1541,6 +1541,88 @@ ldltfact{T<:Real}(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
+    if lF.n != lC.nrow
+        throw(DimensionMismatch("matrix dimensions do not fit"))
+    end
+    @isok ccall((:cholmod_l_updown, :libcholmod), Cint,
+        (Cint, Ptr{C_Sparse{Tv}}, Ptr{C_Factor{Tv}}, Ptr{Void}),
+        update, get(C.p), get(F.p), common())
+    F
+end
+
+#Help functions for rank updates
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: helper functions

File: base/sparse/cholmod.jl [hunk 5]
```diff
@@ -1541,6 +1541,88 @@ ldltfact{T<:Real}(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
+function lowrankdowndate!{Tv<:VTypes}(F::Factor{Tv}, V::AbstractArray{Tv})
+    #Reorder and copy V to account for permutation
+    C = lowrank_reorder(V, get_perm(F))
+    lowrankupdowndate!(F, C, Cint(0))
+end
+
+"""
+    lowrankupdate(F::Factor, C) -> FF::Factor
+
+Get an LDLt Factorization of `A+C*C'` given an `LDLt` or `LLt` factorization `F` of `A`.
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: highlight `LDLt` (also in the next docstring)

File: test/sparse/cholmod.jl [hunk 6]
```diff
@@ -702,3 +702,39 @@ end
     @test_throws ArgumentError CHOLMOD.Sparse(A)
 end
 
+#Test sparse low rank update for cholesky decomposion
+A = SparseMatrixCSC{Float64,CHOLMOD.SuiteSparse_long}(10, 5, [1,3,6,8,10,13], [6,7,1,2,9,3,5,1,7,6,7,9],
+    [-0.138843, 2.99571, -0.556814, 0.669704, -1.39252, 1.33814,
+    1.02371, -0.502384, 1.10686, 0.262229, -1.6935, 0.525239])
+AtA = A'*A;
+C0 = [1., 2., 0, 0, 0]
+#Test both cholfact and LDLt with and without automatic permulations
```
Comment (tkelman) [inline test/sparse/cholmod.jl]: permutations

**Comment by mfalt** [comment]
Rebased on master and fixed the typos, thanks for the comments.

**Comment by ViralBShah** [comment]
Awesome!

File: base/sparse/cholmod.jl [hunk 7]
```diff
@@ -1541,6 +1541,88 @@ ldltfact(A::Union{SparseMatrixCSC{T},SparseMatrixCSC{Complex{T}},
...
     kws...) where {T<:Real} = ldltfact(Sparse(A); kws...)
 
+## Rank updates
+
+"""
+    lowrankupdowndate!(F::Factor, C::Sparse, update::Cint)
+
+Update an `LDLt` or `LLt` Factorization `F` of `A` to a factorization of `A±C*C'`.
+
+If sparsity preserving factorization is used, i.e. `L*L'*=P*A*P'` then the new
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: is there an extra `*` in this after the `L'` ? spaces around the operators might make some of these easier to read

**Comment by mfalt** [comment]
Fixed it, nice catch @tkelman!

**Pull request merged**